openssl/openssl.go

```diff
@@ -35,21 +42,74 @@ func Init() error {
 	return nil
 }
 
+// providerAvailable looks through provider's digests
+// checking if there is any that matches the pprog query.
+func providerAvailable(pprog *C.char) bool {
+	C.go_openssl_ERR_set_mark()
+	md := C.go_openssl_EVP_MD_fetch(nil, algProve, pprog)
+	C.go_openssl_ERR_pop_to_mark()
+	if md == nil {
+		return false
+	}
+	C.go_openssl_EVP_MD_free(md)
+	return true
+}
+
 // FIPS returns true if OpenSSL is running in FIPS mode, else returns false.
 func FIPS() bool {
-	return C.go_openssl_FIPS_mode() == 1
+	// FIPS_mode is only defined in OpenSSL 1.
+	if C._g_FIPS_mode != nil {
+		return C.go_openssl_FIPS_mode() == 1
+	}
+	if C.go_openssl_EVP_default_properties_is_fips_enabled(nil) == 0 {
+		return false
+	}
+	// EVP_default_properties_is_fips_enabled can return true even if the FIPS provider isn't loaded,
+	// it is only based on the default properties.
+	return providerAvailable(propFipsYes)
 }
 
 // SetFIPS enables or disables FIPS mode.
+//
+// It implements the following provider fallback logic for OpenSSL 3:
+//    - The "fips" provider is loaded if enabled=true and no loaded provider matches "fips=yes".
+//    - The "default" provider is loaded if enabled=false and no loaded provider matches "fips=no".
+// This logic allows advanced users to define their own providers that match "fips=yes" and "fips=no" using the OpenSSL config file.
 func SetFIPS(enabled bool) error {
-	var mode C.int
+	// FIPS_mode_set is only defined in OpenSSL 1.
+	if C._g_FIPS_mode_set != nil {
+		var mode C.int
+		if enabled {
+			mode = C.int(1)
+		} else {
+			mode = C.int(0)
+		}
+		if C.go_openssl_FIPS_mode_set(mode) != 1 {
+			return newOpenSSLError("openssl: FIPS_mode_set")
+		}
+		return nil
+	}
+	var pprop, provName *C.char
```

Comment:
Could this naming be expanded a little for more clarity? I'm not 100% sure I have this right:

```suggestion
	var provProp, provName *C.char
```